### PR TITLE
Fix: Bull crab balance internal accounting

### PR DESCRIPTION
**Pull request by haythem96**
# Task: Bull crab balance internal accounting

**Should be merged after #693**

## Description

- Track bull strategy balance on crab internally instead of using `crab.balanceOf(bull)`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## FE Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ ] I fully understand the user problem this PR is solving
- [ ] I know who the target user is for this PR and have a deep understanding of that user
- [ ] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

**Review by daryakaviani — APPROVED** [review]
lgtm but tbd on whether we need test

**Review by alpinechicken — COMMENTED** [review]
Makes sense but why is it better to have internal?

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 1]
```diff
@@ -85,11 +85,15 @@ contract BullStrategyTestFork is Test {
 
     function testInitialDeposit() public {
         uint256 crabToDeposit = 10e18;
+        uint256 bullCrabBalanceBefore = bullStrategy.getCrabBalance();
 
         vm.startPrank(user1);
         (uint256 wethToLend, uint256 usdcToBorrow) = _deposit(crabToDeposit);
         vm.stopPrank();
 
+        uint256 bullCrabBalanceAfter = bullStrategy.getCrabBalance();
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:94]: should we also assert that this equals the erc20 balances from balanceOf?

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 2]
```diff
@@ -98,16 +102,21 @@ contract BullStrategyTestFork is Test {
...
         uint256 crabToDepositInitially = 10e18;
+        uint256 bullCrabBalanceBefore = bullStrategy.getCrabBalance();
 
         vm.startPrank(user1);
         (uint256 wethToLend, uint256 usdcToBorrow) = _deposit(crabToDepositInitially);
         vm.stopPrank();
 
+        uint256 bullCrabBalanceAfter = bullStrategy.getCrabBalance();
+
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositInitially), bullCrabBalanceBefore);
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:113]: same comment

File: packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol [hunk 3]
```diff
@@ -116,6 +125,9 @@ contract BullStrategyTestFork is Test {
         (uint256 wethToLendSecond, uint256 usdcToBorrowSecond) = _deposit(crabToDepositSecond);
         vm.stopPrank();
 
+        bullCrabBalanceAfter = bullStrategy.getCrabBalance();
+
+        assertEq(bullCrabBalanceAfter.sub(crabToDepositSecond), bullCrabBalanceBefore);
```
Comment (aleone) [inline packages/bull-vault/test/e2e/BullStrategyTestFork.t.sol:130]: same comment

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol [hunk 4]
```diff
@@ -148,7 +148,7 @@ contract FlashBullTestFork is Test {
             wethToLend
         );
         assertEq(
-            IERC20(crabV2).balanceOf(address(bullStrategy)).sub(crabToBeMinted),
+            bullStrategy.getCrabBalance().sub(crabToBeMinted),
```
Comment (aleone) [inline packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:151]: we are comparing the balanceOf to the strategy storage variable compared to the expected crab amt

we should probably compare the balanceOfs and the local variables and ensure they are consistent with eachother, so more assertions to add here

File: packages/bull-vault/test/e2e/FlashBullTestFork.t.sol [hunk 5]
```diff
@@ -275,7 +275,7 @@ contract FlashBullTestFork is Test {
             5
         );
         assertEq(
-            IERC20(crabV2).balanceOf(address(bullStrategy)).sub(crabToBeMinted),
+            bullStrategy.getCrabBalance().sub(crabToBeMinted),
```
Comment (aleone) [inline packages/bull-vault/test/e2e/FlashBullTestFork.t.sol:278]: same here

**Review by aleone — COMMENTED** [review]
few extra assertions / changes to assertions need to be done

**Pull request merged**